src/build/HHAPIDocBuildStep.php

```diff
@@ -107,14 +101,46 @@ public function buildAll(): void {
       ),
     );
 
+    // HHApiDoc index; needed so that e.g. `File\WriteHandle`'s documentation
+    // generation knows about `IO\WriteHandle`'s methods
+    Log::i("\nCreating cross-reference index");
+    $hh_apidoc_index = shape(
+      'types' => dict[],
+      'newtypes' => dict[],
+      'functions' => dict[],
+      'classes' => dict[],
+      'interfaces' => dict[],
+      'traits' => dict[],
+    );
+    $all_documentables =
+      Vec\flatten(vec[$builtin_defs, $hsl_defs, $hsl_experimental_defs]);
+    foreach ($all_documentables as $documentable) {
+      $def = $documentable['definition'];
+      // types and newtypes are not currently supported by docs.hhvm.com
+      if ($def is ScannedFunction) {
+        $hh_apidoc_index['functions'][$def->getName()] = $documentable;
+        continue;
+      }
+      if ($def is ScannedClass) {
+        $hh_apidoc_index['classes'][$def->getName()] = $documentable;
+        continue;
+      }
+      if ($def is ScannedInterface) {
+        $hh_apidoc_index['interfaces'][$def->getName()] = $documentable;
+        continue;
+      }
+      if ($def is ScannedTrait) {
+        $hh_apidoc_index['traits'][$def->getName()] = $documentable;
+        continue;
+      }
+    }
```

Comment:
I would use `Dict\group_by` to avoid mixing up of functional programming style code and the mutable loop.

Reply:
`Dict\group_by()` doesn't produce shapes, and would need a call to reflection (or `get_class()` to significantly clean this up

Reply:
There also currently isn't functional style here?

Reply:
I think it would be good if we have a high level API in `Facebook\HHAPIDoc`

Reply:
HHAPIDoc does have one; docs.hhvm.com uses its' own as there's multiple sources (hhvm, HHI, hsl, hsl-experimental) and they can contain contradictory definitions (hhvm + hhi) which need resolving.